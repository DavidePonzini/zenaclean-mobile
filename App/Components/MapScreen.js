import React from 'react'
import MapView, { Marker, Callout, Circle } from 'react-native-maps'
import { Button } from 'native-base'
import { ActivityIndicator, Alert, StyleSheet, Text, View } from 'react-native'
import { StackActions, NavigationActions } from 'react-navigation'
import ActionButton from 'react-native-action-button'
import api from '../Services/ApiService'
import geolocationService from '../Services/GeolocationService'
import Fonts from '../Themes/Fonts'
import DateParser from '../Utils/DateParser'
import Icon from 'react-native-vector-icons/FontAwesome'
import Colors from '../Themes/Colors'
import Images from '../Themes/Images'
import Metrics from '../Themes/Metrics'
import { GooglePlacesAutocomplete } from 'react-native-google-places-autocomplete'
import Secrets from 'react-native-config'

export default class MapScreen extends React.Component {
  constructor (props) {
    super(props)
    this.region = geolocationService.getCurrentRegion() // region in the object holds the region we're currently viewing
    this.balance = ''
    this.state = {
      region: this.region, // region in the state holds the last "searched" region
      inserting: false,
      fetching: false,
      markers: [],
      geolocating: false,
      geolocation: null,
      refreshShowing: false,
      zoomShowing: false
    }
    this.logged = api.isLoggedIn()
    this.props.navigation.addListener('didFocus', this.refreshOnFocus)
    this.props.navigation.addListener('willBlur', () => { this.setState({ refreshShowing: false }) })
    this.markersRefs = {}
  }

  static navigationOptions = {
    tabBarLabel: 'Mappa'
  }

  onRegionChanged = (region) => {
    // prompt the user to refresh the markers only when moving or zooming out
    // -- however, if the zoom is TOO MUCH out, don't
    if (region.longitudeDelta > 0.11) {
      this.setState({ refreshShowing: false, zoomShowing: true })
    } else {
      let newState = { zoomShowing: false, moving: false }
      if ((region.latitude - this.state.region.latitude) ** 2 > 0.001 || (region.longitude - this.state.region.longitude) ** 2 > 0.001 ||
        region.longitudeDelta - this.state.region.longitudeDelta > 0.01) {
        newState.refreshShowing = true
      }
      this.setState(newState)
    }
    if (this.state.inserting) {
      this.insertingMarker.animateMarkerToCoordinate(region)
    }
    this.region = region
    geolocationService.setCurrentRegion(region)
  }

  centerOn (region) {
    this.map.animateToRegion(region)
  }

  markerRegionUpdate = () => {
    let latDelta
    let lngDelta
    if (this.region.longitudeDelta > 0.11) {
      latDelta = (this.region.latitudeDelta * 0.11) / this.region.longitudeDelta
      lngDelta = 0.11
    } else {
      latDelta = this.region.latitudeDelta
      lngDelta = this.region.longitudeDelta
    }
    const neLat = this.region.latitude + latDelta / 2
    const swLat = this.region.latitude - latDelta / 2
    const swLng = this.region.longitude - lngDelta / 2
    const neLng = this.region.longitude + lngDelta / 2
    const that = this

    return new Promise(resolve => {
      that.setState({ fetching: true, region: that.region }, () => {
        api.getMarkers(neLat, swLat, swLng, neLng, (res) => { that.setState({ markers: res, refreshShowing: false, fetching: false }, () => resolve()) })
      })
    })
  }

  componentWillMount () {
    return this.markerRegionUpdate()
  }

  navigateToAddReport = () => {
    return this.props.navigation.navigate('AddReport', {
      lat: this.region.latitude,
      lng: this.region.longitude
    })
  }

  navigateToSignUp = () => {
    this.props.navigation.dispatch(StackActions.reset({
      index: 1,
      actions: [
        NavigationActions.navigate({ routeName: 'Main' }),
        NavigationActions.navigate({ routeName: 'SignUp' })
      ]
    }))
  }

  beginMarkerPlacement = () => {
    if (this.logged) {
      this.setState({ inserting: true, region: this.region })
    } else {
      const that = this
      Alert.alert('Attenzione!', 'Registrati per effettuare una segnalazione', [
        {
          text: 'OK',
          onPress: that.navigateToSignUp
        },
        { text: 'PiÃ¹ tardi', style: 'cancel' }
      ])
    }
  }

  navigateToSingleReport = (marker) => {
    return this.props.navigation.navigate('SingleReport', { marker })
  }

  cancelMarkerPlacement = () => {
    this.setState({ inserting: false })
  }

  renderMarkers = () => {
    if (!this.state.inserting) {
      this.markersRefs = {}
      return this.state.markers.map(this.renderMarker)
    }
  }

  renderMarker = (marker, key) => {
    const { date, time } = DateParser.timestampToItalianDate(marker.timestamp)
    const description = marker.description.length > 80 ? marker.description.substring(0, 80) + '...' : marker.description
    return (
      <Marker
        key={key}
        coordinate={{ latitude: marker.latitude, longitude: marker.longitude }}
        ref={(ref) => { this.markersRefs[marker._id] = ref }}
      >
        <Callout style={styles.callout} onPress={() => { this.navigateToSingleReport(marker) }}>
          <Text style={styles.calloutTitle}>{marker.title}</Text>
          <View style={styles.row}>
            <View style={styles.innerContainer}>
              <Text style={styles.address}>{marker.address == null ? 'Indirizzo sconosciuto' : marker.address}</Text>
              <Text style={styles.calloutDescr}>{description}</Text>
              <Text style={styles.timestamp}>{date + ' alle ' + time}</Text>
            </View>
          </View>
        </Callout>
      </Marker>)
  }

  zoomIn = () => {
    this.map.animateToRegion({
      latitude: this.region.latitude,
      longitude: this.region.longitude,
      latitudeDelta: 0.1,
      longitudeDelta: 0.1
    })
  }

  geolocateMe = () => {
    const that = this
    geolocationService.geolocateOnce().then(pos => {
      that.setState({ geolocating: true, geolocation: { latitude: pos.coords.latitude, longitude: pos.coords.longitude, radius: pos.coords.accuracy } })
      const newRegion = {
        latitude: pos.coords.latitude,
        longitude: pos.coords.longitude,
        longitudeDelta: 0.01,
        latitudeDelta: 0.01
      }
      that.region = newRegion
      that.map.animateToRegion(newRegion, 200)
      setTimeout(that.markerRegionUpdate, 200)
    })
  }
  refreshOnFocus = (payload) => {
    this.markerRegionUpdate().then(() => {
      if (payload.state.params.markerId != null) {
        const setParamsAction = NavigationActions.setParams({
          params: { marker: null },
          key: 'Map'
        })
        this.props.navigation.dispatch(setParamsAction)
        this.highlightMarker(payload.state.params.markerId)
      }
    })
  }

  highlightMarker = (id) => {
    const marker = this.markersRefs[id]
    if (marker != null) {
      setTimeout(() => { marker.showCallout() }, 0)
    }
  }

  render () {
    return (
      <View style={styles.map}>
        <MapView
          style={styles.map}
          initialRegion={this.state.region}
          showsMyLocationButton={false}
          loadingEnabled
          pitchEnabled={false}
          toolbarEnabled={false}
          onRegionChangeComplete={this.onRegionChanged}
          ref={(map) => { this.map = map }}
        >
          {this.state.geolocating &&
            <View>
              <Marker anchor={{ x: 0.5, y: 0.5 }} image={Images.locationDot} coordinate={{ latitude: this.state.geolocation.latitude, longitude: this.state.geolocation.longitude }} />
              <Circle strokeColor='lightblue' fillColor='lightblue' center={{ latitude: this.state.geolocation.latitude, longitude: this.state.geolocation.longitude }} radius={this.state.geolocation.radius} />
            </View>
          }
          {this.renderMarkers()}
          {this.state.inserting &&
          <Marker
            ref={ref => { this.insertingMarker = ref }}
            coordinate={this.state.region}
          />
          }
        </MapView>
        <GooglePlacesAutocomplete
          placeholder='Cerca su Google Maps'
          minLength={2}
          autoFocus={false}
          returnKeyType='search'
          listViewDisplayed='false'
          fetchDetails
          renderDescription={row => row.description}
          onPress={(data, details = null) => {
            const region = {
              latitude: details.geometry.location.lat,
              longitude: details.geometry.location.lng,
              latitudeDelta: 0.001,
              longitudeDelta: 0.001
            }
            this.centerOn(region)
          }}
          getDefaultValue={() => ''}
          query={{
            key: Secrets.GOOGLE_MAPS_API_KEY,
            language: 'it'
          }}
          styles={{
            textInputContainer: {
              width: '100%'
            },
            description: {
              fontWeight: 'bold',
              color: 'black'
            },
            listView: {
              backgroundColor: '#ffffff'
            }
          }}
          debounce={200}
        />
        { !this.state.inserting && this.state.refreshShowing &&
        <Button
          accessibilityLabel='update-markers-button'
          testID={'update-markers-button'}
          style={styles.button_update_markers}
          onPress={this.markerRegionUpdate}
          iconLeft>
          <View style={{ flexDirection: 'row', paddingLeft: 10, paddingRight: 10 }}>
            { (this.state.fetching && <ActivityIndicator style={{ flex: 0.2 }} />) ||
            <Icon name='search' style={{ fontSize: 20, flex: 0.2 }} /> }
            <Text style={{ fontWeight: 'bold', flex: 0.8 }}>Cerca in quest'area</Text>
          </View>
        </Button>}
        { !this.state.inserting && this.state.zoomShowing &&
        <Button
          accessibilityLabel='zoom-button'
          testID={'zoom-button'}
          style={styles.button_update_markers}
          onPress={this.zoomIn}
          iconLeft>
          <View style={{ flexDirection: 'row', paddingLeft: 10, paddingRight: 10 }}>
            <Icon name='search-plus' style={{ fontSize: 20, flex: 0.2 }} />
            <Text style={{ fontWeight: 'bold', fontSize: 12, flex: 0.8 }}>Ingrandisci per cercare</Text>
          </View>
        </Button>}
        <ActionButton
          fixNativeFeedbackRadius
          accessibilityLabel='geolocate-button'
          testID={'geolocate-button'}
          buttonColor='#FAFAFA'
          offsetY={50}
          offsetX={-25}
          position='left/center'
          onPress={this.geolocateMe}
          renderIcon={() => (<Icon name='crosshairs' style={{ fontSize: 20, color: this.state.geolocating ? Colors.accent : Colors.inactiveIcon }} />)}
        />
        <ActionButton fixNativeFeedbackRadius backgroundTappable
          accessibilityLabel='button-add'
          testID={'button-add'}
          buttonColor={Colors.accent}
          offsetY={50}
          onPress={this.beginMarkerPlacement}
          onReset={this.cancelMarkerPlacement}
        >
          {this.logged &&
          <ActionButton.Item accessibilityLabel='button-confirm' testID={'button-confirm'}
            buttonColor={Colors.accent} title='Scegli questa posizione'
            onPress={this.navigateToAddReport}>
            <Icon name='check' style={{ color: 'white' }} />
          </ActionButton.Item>
          }
        </ActionButton>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  calloutTitle: {
    ...Fonts.style.h6,
    borderBottomColor: '#EAEAEA',
    textAlign: 'center',
    paddingBottom: 2.5,
    marginBottom: 5,
    borderBottomWidth: 1
  },
  searchBarStyle: {
    backgroundColor: 'white',
    opacity: 0.8,
    marginTop: 15,
    marginLeft: 10,
    marginRight: 10
  },
  textSearchBar: {
    color: '#848484'
  },
  calloutDescr: {
    ...Fonts.style.description,
    flex: 0.8
  },
  timestamp: {
    ...Fonts.style.footer,
    flex: 0.2
  },
  geoBtn: {
    position: 'absolute',
    left: 10,
    bottom: 50,
    justifyContent: 'center',
    backgroundColor: '#FAFAFA',
    width: 50
  },
  button_update_markers: {
    position: 'absolute',
    top: 100,
    width: 200,
    right: Metrics.width / 2 - 100,
    justifyContent: 'center',
    backgroundColor: Colors.materialGray,
    color: Colors.defaultText
  },
  address: {
    fontSize: Fonts.size.small,
    flex: 0.2
  },
  row: {
    flexDirection: 'row',
    flex: 1,
    padding: 2.5
  },
  innerContainer: {
    flexDirection: 'column',
    flex: 1
  },
  map: {
    ...StyleSheet.absoluteFillObject
  },
  callout: {
    width: 250,
    height: 150
  },
  floatingMarker: {
    position: 'absolute',
    zIndex: 10,
    left: Metrics.width / 2,
    top: Metrics.height / 2,
    fontSize: 50
  }
})
