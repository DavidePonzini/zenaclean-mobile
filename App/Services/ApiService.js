import Secrets from 'react-native-config'

const baseUrl = Secrets.API_URL

const getAddressFromCoords = ({ lat, lng }, cb) => {
  const googleApiUrl = 'https://maps.googleapis.com/maps/api/geocode/json?address='

  fetch(googleApiUrl + lat + ',' + lng + '&key=' + Secrets.GOOGLE_MAPS_API_KEY)
    .then((response) => response.json())
    .then((responseJson) => cb(responseJson))
}

const getMarkers = (cb) => {
  fetch(baseUrl + 'markers.json')
    .then((response) => response.json())
    .then((responseJson) => { cb(responseJson) })
}

const uploadReport = (infoReport) => {
  return fetch(baseUrl + 'markers.json', {
    method: 'POST',
    body: JSON.stringify(infoReport)
  })
}
export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport
}
