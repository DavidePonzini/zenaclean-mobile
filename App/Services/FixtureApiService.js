const SSNlist = ['LNGSVV60T17H509N', 'HTHHLG90R62B112N', 'BRTHZG56P68C982U']
const emailList = ['prova@gmail.com', 'prova@hotmail.it']
export const user1 = {
  email: emailList[0],
  password: '12345678',
  ssn: SSNlist[0],
  id: '5c5afd7c4be7ff1c8d7daacd'
}

const wallets = {}
wallets[user1.id] = 2
export { wallets }
export const fixtureAddress = 'Via del Camoscio'
export const fixtureMarkers = [{ address: fixtureAddress, title: 'Divano abbandonato', latitude: 44.1234, longitude: 8.1234, description: 'Divano abbandonato in via Gramsci', timestamp: '2018-12-09T14:40:35 -02:00', approved_positive: false, approved_negative: false, _id: '5c6580a33a7f647afccd8920', user_id: user1.id, url: 'images/667856780e39f0bbe7407bfdae603f8e', voted_positive: false, voted_negative: false }]
export const fixtureMessageOk = 'Registrazione effettuata con successo!'
export const fixtureMessageNotOk = 'Utente giÃ  esistente'
export const fixtureUserId = 'verylonguserid'
const totVoti = [32, 45, 21]
export let userId = null

const logInUser = (email, password, cb) => {
  return new Promise(resolve => {
    (user1.email === email && user1.password === password) ? resolve({ status: 'ok' }) : resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      userId = fixtureUserId
      cb(null, { id: userId, email, eth_address: null })
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const getAddressFromCoords = ({ lat, lng }, cb) => {
  return new Promise((resolve) => {
    resolve(fixtureAddress)
  }).then((res) => { cb(null, res) })
}

const getMarkers = (neLat, swLat, swLng, neLng, cb) => {
  return new Promise((resolve) => {
    resolve(JSON.stringify(fixtureMarkers))
  }).then((res) => { cb(JSON.parse(res)) })
}

const uploadReport = (infoReport, cb) => {
  return new Promise((resolve) => {
    resolve({ status: 'ok' })
  }).then(() => {
    cb(null)
  })
}

const registerUser = (email, ssn, password, cb) => {
  return new Promise((resolve) => {
    let checked = false
    for (let i = 0; i < SSNlist.length; i++) {
      if (SSNlist[i] === ssn) checked = true
    }
    for (let i = 0; i < emailList.length; i++) {
      if (emailList[i] === email) checked = true
    }
    if (!checked) { resolve(fixtureMessageOk) }
    if (checked) { resolve(fixtureMessageNotOk) }
  }).then((res) => { cb(null, res) })
}

const voteReport = (v, report, cb) => {
  return new Promise((resolve) => {
    let old = totVoti[report]
    let current = totVoti[report] + 1
    if (old < current) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res, current) => {
    if (res.status === 'ok') {
      totVoti[report] = current
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const changePassword = (email, oldPassword, newPassword, confirmPassword, cb) => {
  return new Promise((resolve) => {
    if (user1.password === oldPassword && newPassword === confirmPassword) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      user1.password = newPassword
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const isLoggedIn = () => userId != null

const logoutUser = () => {
  userId = null
}
const rehydrateLogin = () => {
  return new Promise((resolve, reject) => {
    if (userId != null) {
      resolve(userId)
    } else {
      reject(new Error('Not logged in'))
    }
  })
}

const getBalance = (eth_addr, cb) => {
  return new Promise(resolve => {
    resolve({ value: wallets[user1.id] })
  }).then(res => {
    cb(res.value)
  })
}

export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport,
  registerUser,
  logInUser,
  changePassword,
  voteReport,
  isLoggedIn,
  logoutUser,
  rehydrateLogin,
  getBalance
}
