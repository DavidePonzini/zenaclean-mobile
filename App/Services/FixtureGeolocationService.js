const fixturePosition = {
  coords: {
    accuracy: 1135,
    altitude: null,
    altitudeAccuracy: null,
    heading: null,
    latitude: 44.4135423,
    longitude: 8.9554223,
    speed: null
  },
  timestamp: 1547641567823
}

const fixtureRegion = {
  latitude: 44.4135423,
  longitude: 8.9554223,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

let lastLocation = fixturePosition
let currentRegion = fixtureRegion

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    let pos = fixturePosition
    lastLocation = pos
    resolve(pos)
  })
}

const getLatestPosition = () => lastLocation

const geolocateContinuously = (cb, cbErr, options) => {
  return setInterval(cb, 100)
}

const stopGeolocating = (id) => {
  clearInterval(id)
}

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
