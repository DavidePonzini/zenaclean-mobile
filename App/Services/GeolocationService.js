let lastLocation = null

// initial region
let currentRegion = {
  latitude: 44.4056,
  longitude: 8.9463,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    navigator.geolocation.getCurrentPosition((pos) => {
      lastLocation = pos
      resolve(pos)
    })
  })
}

const geolocateContinuously = (cb, cbErr, options) => {
  return navigator.geolocation.watchPosition((pos) => {
    cb(pos)
    lastLocation = pos
  }, cbErr, options)
}

const stopGeolocating = (id) => {
  navigator.geolocation.clearWatch(id)
}

const getLatestPosition = () => lastLocation

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
