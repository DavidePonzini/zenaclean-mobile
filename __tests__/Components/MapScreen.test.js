import React from 'react'
import MapScreen from '../../App/Components/MapScreen'

import { fixtureMarkers } from '../../App/Services/FixtureApiService'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import 'isomorphic-fetch'
import { shallowToJson } from 'enzyme-to-json'
import { Marker, Callout } from 'react-native-maps'

configure({ adapter: new Adapter() })

jest.mock('../../App/Services/ApiService', () => require('../../App/Services/FixtureApiService'))

const navigation = { navigate: jest.fn(), state: { params: { logged: true } } }

describe('MapScreen tests', () => {
  const wrapper = shallow(<MapScreen navigation={navigation} />)
  it('renders correctly', async () => {
    const instance = wrapper.instance()
    await instance.componentWillMount()
    wrapper.update()
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('downloads markers', async () => {
    const instance = wrapper.instance()
    await instance.componentWillMount()
    expect(instance.state.markers.length).toBe(fixtureMarkers.length)
  })

  it('renders markers', () => {
    const instance = wrapper.instance()
    const marker = instance.renderMarker(fixtureMarkers[0])
    expect(marker.type).toBe(Marker)
  })

  it('navigates when marker callout is pressed', async () => {
    const instance = wrapper.instance()
    await instance.componentWillMount()
    const marker = wrapper.find(Marker).first()
    marker.simulate('Press')
    wrapper.update()
    const callout = wrapper.find(Callout).first()
    callout.simulate('Press')
    expect(navigation.navigate).toHaveBeenCalled()
  })
})
