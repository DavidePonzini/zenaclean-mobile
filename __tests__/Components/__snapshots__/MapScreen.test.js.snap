// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapScreen tests renders correctly 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <MapView
    initialRegion={
      Object {
        "latitude": 44.4056,
        "latitudeDelta": 0.1,
        "longitude": 8.9463,
        "longitudeDelta": 0.1,
      }
    }
    onRegionChange={[Function]}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <MapMarker
      coordinate={
        Object {
          "latitude": 44.1234,
          "longitude": 8.1234,
        }
      }
      key="0"
      stopPropagation={false}
    >
      <MapCallout
        onPress={[Function]}
        style={
          Object {
            "height": 150,
            "width": 250,
          }
        }
        tooltip={false}
      >
        <Text
          style={
            Object {
              "borderBottomColor": "#EAEAEA",
              "borderBottomWidth": 1,
              "fontFamily": "HelveticaNeue-Italic",
              "fontSize": 19,
              "marginBottom": 5,
              "paddingBottom": 2.5,
              "textAlign": "center",
            }
          }
        >
          Divano abbandonato
        </Text>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "padding": 2.5,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
              }
            }
          >
            <Text
              style={
                Object {
                  "flex": 0.2,
                  "fontSize": 12,
                }
              }
            >
              Via del Camoscio
            </Text>
            <Text
              style={
                Object {
                  "flex": 0.8,
                  "fontFamily": "Avenir-Book",
                  "fontSize": 14,
                }
              }
            >
              Divano abbandonato in via Gramsci
            </Text>
            <Text
              style={
                Object {
                  "color": "#808080",
                  "flex": 0.2,
                  "fontFamily": "Avenir-Book",
                  "fontSize": 12,
                }
              }
            >
              9 Dicembre 2018 alle 17:40
            </Text>
          </View>
        </View>
      </MapCallout>
    </MapMarker>
  </MapView>
  <ActionButton
    active={false}
    activeOpacity={0.85}
    autoInactive={true}
    backdrop={false}
    backgroundTappable={true}
    bgColor="transparent"
    bgOpacity={1}
    buttonColor="dodgerblue"
    buttonText="+"
    buttonTextStyle={Object {}}
    degrees={45}
    fixNativeFeedbackRadius={true}
    nativeFeedbackRippleColor="rgba(255,255,255,0.75)"
    offsetX={30}
    offsetY={50}
    onPress={[Function]}
    onPressIn={[Function]}
    onPressOn={[Function]}
    onReset={[Function]}
    outRangeScale={1}
    position="right"
    resetToken={null}
    size={56}
    spacing={20}
    useNativeFeedback={true}
    verticalOrientation="up"
  >
    <ActionButtonItem
      active={true}
      activeOpacity={0.85}
      buttonColor="dodgerblue"
      fixNativeFeedbackRadius={false}
      nativeFeedbackRippleColor="rgba(255,255,255,0.75)"
      onPress={[Function]}
      spaceBetween={15}
      title="Scegli questa posizione"
      useNativeFeedback={true}
    >
      <Icon
        allowFontScaling={false}
        name="check"
        size={12}
        style={
          Object {
            "color": "white",
          }
        }
      />
    </ActionButtonItem>
  </ActionButton>
</View>
`;
